### PR TITLE
Complete documentation rework

### PR DESCRIPTION
**Pull request by tgraf**
``` release-note
Complete rework of documentation including policy examples
```

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -9,51 +9,131 @@
...
-networking, security, loadbalancing and visibility.
+connectivity and loadbalancing between application workloads such as
+application containers or processes. Cilium operates at Layer 3/4 to provide
+traditional networking and security services as well as Layer 7 to protect and
+secure use of modern application protocols such as HTTP, gRPC and Kafka. Cilium
+is integrates to common orchestration frameworks such as Kubernetes and Mesos.
+
+A new Linux kernel technology called BPF is at the foundation of Cilium. It
+supports dynamic insertion of BPF bytecode into the Linux kernel at various
+integration points such as: network IO, application sockets, and traceptions to
```
Comment (borkmann) [inline README.md]: `traceptions`

File: README.md [hunk 2]
```diff
@@ -9,51 +9,131 @@
...
-
-A new Linux kernel technology called eBPF is at the foundation of Cilium, which
-enables the dynamic insertion of BPF bytecode into the Linux kernel. Cilium
-generates eBPF programs for each individual application container to provide
-networking, security, loadbalancing and visibility.
+connectivity and loadbalancing between application workloads such as
+application containers or processes. Cilium operates at Layer 3/4 to provide
+traditional networking and security services as well as Layer 7 to protect and
+secure use of modern application protocols such as HTTP, gRPC and Kafka. Cilium
+is integrates to common orchestration frameworks such as Kubernetes and Mesos.
```
Comment (borkmann) [inline README.md]: `is integrates`

File: README.md [hunk 3]
```diff
@@ -9,51 +9,131 @@
...
+
+A new Linux kernel technology called BPF is at the foundation of Cilium. It
+supports dynamic insertion of BPF bytecode into the Linux kernel at various
+integration points such as: network IO, application sockets, and traceptions to
+implement security, networking and visibility logic. BPF is highly efficient
+and flexible. To learn more about BPF, read mode in our our extensive [BPF
```
Comment (borkmann) [inline README.md]: `our our`

File: README.md [hunk 4]
```diff
@@ -9,51 +9,131 @@
...
-   Kubernetes services. North-south traffic to load balance external traffic,
-   e.g. implementation of Kubernetes ingress. All load-balancing performed
-   with direct server return (DSR) by default for improved performance.
- * **Troubleshooting:** Built-in troubleshooting tools providing an alternative
-   to traditional tcpdump troubleshooting techniques.
+## Functionality Overview
+
+ * **Protect and secure APIs transparently:** Ability to secure modern
+   application protocols such as REST/HTTP, gRPC and Kafka. Traditional
+   firewalls operates at Layer 3 and 4. A protocol running on a particular port
```
Comment (borkmann) [inline README.md]: `operates`

File: README.md [hunk 5]
```diff
@@ -9,51 +9,131 @@
...
+
+ * **Simple Networking:** A simple flat Layer 3 network with the ability to
+   span multiple clusters connects all application containers. IP allocation is
+   kept simple by using host scope allocators. This means that each host can
+   allocate IPs without any coordination between hosts.
+
+   The following multi node networking models are supported:
+
+   * **Overlay:** Encapsulation based virtual network spawning all hosts.
+     Currently VXLAN and Geneve are baked in but all encapsulation formats
```
Comment (borkmann) [inline README.md]: `baked` .. maybe : `are supported natively`

File: README.md [hunk 6]
```diff
@@ -9,51 +9,131 @@
...
+     - Native IPv6 networks
+     - In conjunction with cloud network routers
+     - If you are already running routing daemons
+
+   Additional transport mechanisms will be supported in the future, see the
+   [roadmap][roadmap].
+
+ * **Load balancing:** Distributed load balancing for traffic between
+   application containers and to external services. The loadbalancing is
+   implemented using BPF using efficient hashtables allowing for almost
```
Comment (borkmann) [inline README.md]: `using BPF using`

File: README.md [hunk 7]
```diff
@@ -9,51 +9,131 @@
...
+
+ * **Monitoring and Troubleshooting:** The ability to gain visibility and to
+   troubleshoot issues is fundamental to the operation of any distributed
+   system. While we learned to love tools like `tcpdump` and `ping` and while
+   they will always find a special place in our hearts, we strive to provide
+   better tooling for troubleshooting. This includes tooling to provide:
+
+   - Event monitoring with metadata: When a packet is dropped, the tool doesn't
+     just report the source and destination IP of the packet, the tool provides
+     the full label information of both the sender and receiving among a lot of
```
Comment (borkmann) [inline README.md]: `among a lot of`

File: README.md [hunk 8]
```diff
@@ -9,51 +9,131 @@
...
+     other information.
+
+   - Policy decision tracing: Why is a packet being dropped or a request
+     rejected. The policy tracing framework allows to trace the policy decision
+     process for both, running workloads and based on artbirary label
```
Comment (borkmann) [inline README.md]: `artbirary `

File: README.md [hunk 9]
```diff
@@ -97,11 +178,15 @@ where programs can be attached to in order to allow for a programmable, high
 performance packet processor in the Linux kernel networking data path.
```
Comment (borkmann) [inline README.md]: Can we also fix the graphics in the README where we have `TC ingress` twice mistakenly?

**Review by borkmann — CHANGES_REQUESTED** [review]

File: README.rst [hunk 10]
```diff
@@ -0,0 +1,307 @@
...
+traditional networking and security services as well as Layer 7 to protect and
+secure use of modern application protocols such as HTTP, gRPC and Kafka. Cilium
+is integrates to common orchestration frameworks such as Kubernetes and Mesos.
+
+A new Linux kernel technology called BPF is at the foundation of Cilium. It
+supports dynamic insertion of BPF bytecode into the Linux kernel at various
+integration points such as: network IO, application sockets, and traceptions to
+implement security, networking and visibility logic. BPF is highly efficient
+and flexible. To learn more about BPF, read mode in our our extensive
+`BPF and XDP Reference Guide`_.
```
Comment (aalemayhu) [inline README.rst]: s/read mode/read more/g

File: README.rst [hunk 11]
```diff
@@ -0,0 +1,307 @@
...
+
+- Event monitoring with metadata: When a packet is dropped, the tool doesn't
+  just report the source and destination IP of the packet, the tool provides
+  the full label information of both the sender and receiving among a lot of
+  other information.
+
+- Policy decision tracing: Why is a packet being dropped or a request rejected.
+  The policy tracing framework allows to trace the policy decision process for
+  both, running workloads and based on artbirary label definitions.
+
```
Comment (aalemayhu) [inline README.rst]: s/artbirary/arbitrary/g

File: README.rst [hunk 12]
```diff
@@ -0,0 +1,307 @@
...
+for native execution efficiency. BPF programs can be run at various hooking
+points in the kernel such as for incoming packets, outgoing packets, system
+calls, kprobes, uprobes, tracepoints, etc.
+
+BPF continues to evolve and gain additional capabilities with each new Linux
+release. Cilium leverages BPF to perform core data path filtering, mangling,
+monitoring and redirection, and requires BPF capabilities that are in any Linux
+kernel version 4.8.0 or newer (the latest current stable Linux kernel is
+4.10.x).
+
```
Comment (aalemayhu) [inline README.rst]: s/4.10.x/4.14.x/g

**Review by aalemayhu — APPROVED** [review]
Just some minor things. Rest looks good. Thanks for fixing the dependencies section, will drop them from my branch.

In one of your commits, you are updating sphinx. Is this process documented somewhere?

**Pull request merged**